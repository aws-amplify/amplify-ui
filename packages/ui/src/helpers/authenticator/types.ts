/**
 * This files provides types that describe general shape of
 * authenticator machine and its intepreter.
 */
import { Interpreter } from 'xstate';
import { AuthUser } from 'aws-amplify/auth';
import { HubCapsule } from 'aws-amplify/utils';

import { AuthContext, AuthEvent } from '../../machines/authenticator/types';

/**
 * Intefrace for `authMachine` machine interpreter
 *
 * TODO: tighten up anys here
 */
export type AuthInterpreter = Interpreter<
  AuthContext,
  any,
  AuthEvent,
  any,
  any
>;

/**
 * Function type for `send` in `authMachine`
 */
export type AuthMachineSend = AuthInterpreter['send'];

/**
 * Handles Amplify JS Auth hub events, by forwarding hub events as appropriate
 * xstate events.
 */
export type AuthMachineHubHandler = (
  data: HubCapsule<any, any>,
  service: AuthInterpreter,
  options?: {
    onSignIn?: (user: AuthUser) => void;
    onSignOut?: () => void;
  }
) => void;
