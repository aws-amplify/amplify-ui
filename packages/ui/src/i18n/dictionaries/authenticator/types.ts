type AuthenticatorText =
  | 'Account recovery requires verified contact information'
  | 'Back to Sign In'
  | 'Change Password'
  | 'Changing'
  | 'Code'
  | 'Confirm Password'
  | 'Confirm Sign Up'
  | 'Confirm SMS Code'
  | 'Confirm TOTP Code'
  | 'Confirm'
  | 'Confirmation Code'
  | 'Confirming'
  | 'Create a new account'
  | 'Create Account'
  | 'Creating Account'
  | 'Dismiss alert'
  | 'Email'
  | 'Enter your code'
  | 'Enter your email'
  | 'Enter your phone number'
  | 'Enter your username'
  | 'Forgot your password?'
  | 'Hide password'
  | 'Loading'
  | 'New password'
  | 'Password'
  | 'Phone Number'
  | 'Resend Code'
  | 'Reset your password'
  | 'Reset your Password'
  | 'Send code'
  | 'Send Code'
  | 'Sending'
  | 'Setup TOTP'
  | 'Show password'
  | 'Sign in to your account'
  | 'Sign In with Amazon'
  | 'Sign In with Apple'
  | 'Sign In with Facebook'
  | 'Sign In with Google'
  | 'Sign in'
  | 'Sign In'
  | 'Signing in'
  | 'Skip'
  | 'Submit'
  | 'Submitting'
  | 'Username'
  | 'Verify Contact'
  | 'Verify';

// Including Record<string, string> to handle additional translations provided by customers
export type AuthenticatorDictionary = Record<AuthenticatorText, string> &
  Record<string, string>;
