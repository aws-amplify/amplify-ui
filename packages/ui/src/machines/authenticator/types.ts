import { State } from 'xstate';
import { AuthUser } from 'aws-amplify/auth';

import {
  LoginMechanism,
  SignUpAttribute,
  SocialProvider,
  UnverifiedUserAttributes,
  AuthFormData,
  AuthFormFields,
  AuthTouchData,
  ValidationError,
  PasswordSettings,
} from '../../types';

import { defaultServices } from './defaultServices';

// copied from JS v6 types
export type ChallengeName =
  | 'SMS_MFA'
  | 'SOFTWARE_TOKEN_MFA'
  | 'SELECT_MFA_TYPE'
  | 'MFA_SETUP'
  | 'PASSWORD_VERIFIER'
  | 'CUSTOM_CHALLENGE'
  | 'DEVICE_SRP_AUTH'
  | 'DEVICE_PASSWORD_VERIFIER'
  | 'ADMIN_NO_SRP_AUTH'
  | 'NEW_PASSWORD_REQUIRED';

/**
 * `AuthDeliveryMedium` is deeply nested in the v6 types, added this as utility
 */
export type V6AuthDeliveryMedium = 'EMAIL' | 'SMS' | 'PHONE' | 'UNKNOWN';

// Utility type representing the `codeDeliveryDetails` from JS v5
export interface V5CodeDeliveryDetails {
  AttributeName: string;
  DeliveryMedium: V6AuthDeliveryMedium;
  Destination: string;
}

// v6 `codeDeliveryDetails`
export interface CodeDeliveryDetails {
  destination?: string;
  deliveryMedium?: V6AuthDeliveryMedium;
  attributName?: string;
}

/**
 * Authenticator routes that can be directly navigated to by user interaction.
 */
export type NavigableRoute = 'signIn' | 'signUp' | 'forgotPassword';
export type InitialRoute = 'signIn' | 'signUp' | 'forgotPassword';

/**
 * `AuthTOTPSetupDetails` is deeply nested in the v6 types, added this as utility
 */
export interface AuthTOTPSetupDetails {
  sharedSecret: string;
  getSetupUri: (appName: string, accountName?: string) => URL;
}

/**
 * Events that occur when actors are done
 */
export type InvokeActorEventTypes =
  | 'done.invoke.forgotPasswordActor'
  | 'done.invoke.signInActor'
  | 'done.invoke.signUpActor'
  | 'done.invoke.signOutActor'
  | 'done.invoke.verifyUserAttributesActor';

/**
 * All known explicit events for xstate
 */
export type AuthEventTypes =
  | 'CHANGE'
  | 'BLUR'
  | 'FEDERATED_SIGN_IN'
  | 'RESEND'
  | 'FORGOT_PASSWORD'
  | 'AUTO_SIGN_IN_FAILURE'
  | 'SIGN_IN_WITH_REDIRECT'
  | 'SIGN_IN'
  | 'SIGN_OUT'
  | 'SIGN_UP'
  | 'SKIP'
  | 'SUBMIT'
  | 'INIT'
  | 'TOKEN_REFRESH'
  | InvokeActorEventTypes;

/**
 * Data payload for auth events
 */
export type AuthEventData = Record<PropertyKey, any>; // TODO: this should be typed further

/** Top-level auth machine event interface */
export interface AuthEvent {
  type: AuthEventTypes;
  data?: AuthEventData;
}

/**
 * Data that actor returns when they are done and reach the final state
 */
export interface ActorDoneData {
  challengeName?: ChallengeName;
  codeDeliveryDetails?: V5CodeDeliveryDetails;
  missingAttributes?: string[];
  remoteError?: string;
  step: Step;
  totpSecretCode?: string;
  username?: string;
  unverifiedUserAttributes?: UnverifiedUserAttributes;
}

/**
 * Context interface for the top-level machine
 */
export interface AuthContext {
  actorRef?: any;
  config?: {
    loginMechanism?: LoginMechanism;
    loginMechanisms?: LoginMechanism[];
    signUpAttributes?: SignUpAttribute[];
    socialProviders?: SocialProvider[];
    formFields?: AuthFormFields;
    initialState?: 'signIn' | 'signUp' | 'forgotPassword';
    passwordSettings?: PasswordSettings;
  };
  services?: Partial<typeof defaultServices>;
  user?: AuthUser;
  // data returned from actors when they finish and reach their final state
  actorDoneData?: ActorDoneData;
  hasSetup?: boolean;
}

// maps to `initialState`
export type InitialStep = 'FORGOT_PASSWORD' | 'SIGN_IN' | 'SIGN_UP';

export type SignInStep =
  | 'CONFIRM_SIGN_IN_WITH_SMS_CODE'
  | 'CONFIRM_SIGN_IN_WITH_TOTP_CODE'
  | 'CONFIRM_SIGN_IN_WITH_NEW_PASSWORD_REQUIRED'
  | 'CONFIRM_SIGN_UP'
  | 'CONTINUE_SIGN_IN_WITH_TOTP_SETUP'
  | 'RESET_PASSWORD'
  | 'SIGN_IN_COMPLETE'; // 'DONE'

export type ResetPasswordStep =
  | 'CONFIRM_RESET_PASSWORD_WITH_CODE'
  | 'RESET_PASSWORD_COMPLETE'; // 'DONE'

export type SignUpStep =
  | 'CONFIRM_SIGN_UP'
  | 'COMPLETE_AUTO_SIGN_IN'
  | 'SIGN_UP_COMPLETE'; // 'DONE'

export type UserAttributeStep =
  | 'SHOULD_CONFIRM_USER_ATTRIBUTE'
  | 'CONFIRM_ATTRIBUTE_WITH_CODE'
  | 'CONFIRM_ATTRIBUTE_COMPLETE'; // 'DONE'

type Step =
  | InitialStep
  | SignInStep
  | SignUpStep
  | ResetPasswordStep
  | UserAttributeStep;

/**
 * Base context for all actors that have auth forms associated
 */
interface BaseFormContext {
  // key/values derived directly from Auth API return values
  challengeName?: ChallengeName;
  missingAttributes?: Array<string>;
  remoteError?: string;
  step: Step;
  totpSecretCode?: string;
  unverifiedUserAttributes?: UnverifiedUserAttributes;

  // kept in memory for submission to relevnat APIs
  username?: string;
  selectedUserAttribute?: string;

  // retrieved from the Auth module on sign in,
  // cleared on sign out
  user?: AuthUser;

  // configuration override key/values
  loginMechanisms: Required<AuthContext>['config']['loginMechanisms'];
  passwordSettings?: PasswordSettings;
  socialProviders: Required<AuthContext>['config']['socialProviders'];
  signUpAttributes?: SignUpAttribute[];

  // form state key/values
  formFields?: AuthFormFields;
  formValues?: AuthFormData;
  touched?: AuthTouchData;
  validationError?: ValidationError;
}

export interface ResetPasswordContext extends BaseFormContext, ActorDoneData {}
export interface SignInContext extends BaseFormContext, ActorDoneData {}
export interface SignUpContext extends BaseFormContext, ActorDoneData {}
export interface VerifyUserContext extends BaseFormContext, ActorDoneData {}

export interface SignOutContext extends Pick<BaseFormContext, 'user'> {}

export type AuthActorContext =
  | SignInContext
  | SignUpContext
  | ResetPasswordContext;

// Actor States
export type SignInState = State<SignInContext, AuthEvent>;
export type SignUpState = State<SignUpContext, AuthEvent>;
export type SignOutState = State<SignOutContext, AuthEvent>;
export type ResetPasswordState = State<ResetPasswordContext, AuthEvent>;
export type AuthActorState = State<AuthActorContext, AuthEvent>;

// Top level machine state
export type AuthMachineState = State<AuthContext, AuthEvent>;
