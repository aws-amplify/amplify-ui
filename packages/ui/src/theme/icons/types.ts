type ViewBox = {
  minX?: number;
  minY?: number;
  width?: number;
  height?: number;
};

export type Icon = {
  viewBox?: ViewBox;
  pathData?: string;
  paths?: React.SVGAttributes<SVGPathElement>[];
};
