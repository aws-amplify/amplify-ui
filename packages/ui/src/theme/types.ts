import { PartialDeep } from '../types';
import { DefaultTokens, Tokens, WebTokens } from './tokens';
import { Breakpoints } from './breakpoints';

export * from './tokens/types/designToken';
export type { BorderWidths } from './tokens/borderWidths';
export type { FontSizes } from './tokens/fontSizes';
export type { FontWeights } from './tokens/fontWeights';
export type { LineHeights } from './tokens/lineHeights';
export type { Radii } from './tokens/radii';
export type { Shadows } from './tokens/shadows';
export type { SpaceSizes } from './tokens/space';

/**
 * An override is a set of tokens that override others
 * in certain contexts. On Android, these are like resource
 * qualifiers. On the web, these are like media queries with
 * a set of CSS variables in it. A theme should contain
 * everything it needs to render
 */
export type Override =
  | SelectorOverride
  | MediaQueryOverride
  | BreakpointOverride
  | ColorModeOverride;

interface BaseOverride {
  tokens?: PartialDeep<Tokens>;
}

/**
 * This override takes a breakpoint name and creates a media-query for that
 * breakpoint
 * ```css
 * @media (min-width: 20em) {
 *    [data-amplify-theme] {
 *      --amplify-font-size-large: 2rem;
 *    }
 * }
 * ```
 */
interface BreakpointOverride extends BaseOverride {
  breakpoint: keyof Breakpoints['values'];
}

/**
 * ```css
 * @media (prefers-color-scheme: dark) {
 *    --amplify-colors-background-primary: black;
 *    --amplify-colors-font-primary: white;
 * }
 * ```
 */
export interface MediaQueryOverride extends BaseOverride {
  mediaQuery: string;
}

/**
 * ```css
 * .disco-theme {
 *    --amplify-colors-background-primary: pink;
 * }
 * [data-my-cool-theme] {
 *    --amplify-colors-font-primary: purple;
 * }
 * ```
 */
export interface SelectorOverride extends BaseOverride {
  selector: string;
}

type ColorMode = 'light' | 'dark';

/**
 * This creates a color mode override, where the color mode is 'light' or 'dark'.
 * Note: there is no 'system' here.
 *
 * ```css
 * @media(prefers-color-scheme: dark) {
 *    [data-amplify-color-mode="system"] {
 *    }
 * }
 *
 * [data-amplify-color-mode="dark"] {
 *
 * }
 * ```
 */
export interface ColorModeOverride extends BaseOverride {
  colorMode: ColorMode;
}

/**
 * A Theme just needs a name. This is what a user would generally deal with.
 * They can define any tokens or breakpoints they need, but they don't need a
 * complete theme with all tokens.
 */
export interface Theme {
  /**
   * The name of the theme. This is used to create scoped CSS to allow for
   * multiple themes on a page.
   */
  name: string;
  tokens?: Tokens;
  breakpoints?: PartialDeep<Breakpoints>;
  /**
   * Overrides allow you to change design tokens in different contexts, like
   * light and dark mode. You can also have other media query overrides as well
   * as breakpoint overrides which correspond to the Breakpoints on a theme,
   * and a generic selector override.
   */
  overrides?: Array<Override>;
}

/**
 * A DefaultTheme has all tokens and breakpoints required
 */
export interface DefaultTheme extends Pick<Theme, 'name' | 'overrides'> {
  tokens: DefaultTokens;
  breakpoints: Breakpoints;
}

/**
 * WebTheme is a fully built theme that has cssText based
 * on the design tokens and all design tokens have added fields
 * to be used in Javascript/Typescript.
 */
export interface WebTheme
  extends Pick<DefaultTheme, 'breakpoints' | 'name' | 'overrides'> {
  cssText: string;
  // property `components` is not specified on `WebTokens`,
  // but is a required token property of `WebTheme`
  tokens: WebTokens;
}
