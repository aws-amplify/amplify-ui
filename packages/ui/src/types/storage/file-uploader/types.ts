export type FileName = string | Array<string>;
