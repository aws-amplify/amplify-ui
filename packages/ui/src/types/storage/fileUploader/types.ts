export type FileName = string | string[];
