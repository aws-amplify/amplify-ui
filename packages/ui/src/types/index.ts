export * from './accountSettings';
export * from './authenticator';
export * from './primitives';
export * from './util';
