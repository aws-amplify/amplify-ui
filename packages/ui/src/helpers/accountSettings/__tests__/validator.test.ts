import { Amplify } from 'aws-amplify';

import {
  getDefaultPasswordValidators,
  getHasMinLength,
  getPasswordRequirement,
  hasLowerCase,
  hasNumber,
  hasSpecialChar,
  hasUpperCase,
  getMatchesConfirmPassword,
  shouldValidate,
  runFieldValidators,
  getDefaultConfirmPasswordValidators,
} from '../validator';
import { ValidatorOptions } from '../../../types';

const partialAmplifyPasswordConfig = {
  aws_cognito_password_protection_settings: {
    passwordPolicyCharacters: ['REQUIRES_LOWERCASE', 'REQUIRES_UPPERCASE'],
  },
};

const fullAmplifyPasswordConfig = {
  aws_cognito_password_protection_settings: {
    passwordPolicyMinLength: 8,
    passwordPolicyCharacters: [
      'REQUIRES_LOWERCASE',
      'REQUIRES_NUMBERS',
      'REQUIRES_SYMBOLS',
      'REQUIRES_UPPERCASE',
    ],
  },
};

const configureSpy = jest.spyOn(Amplify, 'configure');

describe('getHasMinLength', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('returns minLength validator as expected', () => {
    const validator = getHasMinLength(4);
    expect(validator).toMatchObject({
      validator: expect.any(Function),
      message: 'Password must have at least 4 characters',
      validationMode: 'onTouched',
    });
  });

  it('validates to true when password is long enough', () => {
    const { validator } = getHasMinLength(4);
    const isValid = validator('longpassword');
    expect(isValid).toBe(true);
  });

  it('validates to false when password is too short', () => {
    const { validator } = getHasMinLength(4);
    const isValid = validator('hi');
    expect(isValid).toBe(false);
  });
});

describe('hasLowerCase', () => {
  it('validates to true when password has lower case', () => {
    const { validator } = hasLowerCase;
    const isValid = validator('password');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no lower case', () => {
    const { validator } = hasLowerCase;
    const isValid = validator('PASSWORD');
    expect(isValid).toBe(false);
  });
});

describe('hasUpperCase', () => {
  it('validates to true when password has upper case', () => {
    const { validator } = hasUpperCase;
    const isValid = validator('PASSWORD');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no upper case', () => {
    const { validator } = hasUpperCase;
    const isValid = validator('password');
    expect(isValid).toBe(false);
  });
});

describe('hasNumber', () => {
  it('validates to true when password has number', () => {
    const { validator } = hasNumber;
    const isValid = validator('password123');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no number', () => {
    const { validator } = hasNumber;
    const isValid = validator('password');
    expect(isValid).toBe(false);
  });
});

describe('hasSpecialChar', () => {
  it('validates to true when password has special character', () => {
    const { validator } = hasSpecialChar;
    const isValid = validator('password!');
    expect(isValid).toBe(true);
  });

  it('validates to false when password has no special character', () => {
    const { validator } = hasSpecialChar;
    const isValid = validator('password');
    expect(isValid).toBe(false);
  });
});

describe('getPasswordRequirement', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('returns null if Amplify.configure() is empty', () => {
    configureSpy.mockReturnValue({});
    const requirements = getPasswordRequirement();
    expect(requirements).toBeNull();
  });

  it('returns expected password requirements if full amplify config', () => {
    configureSpy.mockReturnValue(fullAmplifyPasswordConfig);
    const requirements = getPasswordRequirement();
    expect(requirements).toMatchObject({
      minLength: 8,
      needsLowerCase: true,
      needsNumber: true,
      needsSpecialChar: true,
      needsUpperCase: true,
    });
  });

  it('returns expected password requirements if partial amplify config', () => {
    configureSpy.mockReturnValue(partialAmplifyPasswordConfig);
    const requirements = getPasswordRequirement();
    expect(requirements).toMatchObject({
      needsLowerCase: true,
      needsNumber: false,
      needsSpecialChar: false,
      needsUpperCase: true,
    });
  });
});

describe('getDefaultPasswordValidators', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('returns validators as expected for partial Amplify config ', () => {
    configureSpy.mockReturnValue(partialAmplifyPasswordConfig);
    const validators = getDefaultPasswordValidators();
    expect(validators).toMatchSnapshot();
  });

  it('returns validators as expected for full Amplify config ', () => {
    configureSpy.mockReturnValue(fullAmplifyPasswordConfig);
    const validators = getDefaultPasswordValidators();
    expect(validators).toMatchSnapshot();
  });

  it('returns empty array for empty amplify config', () => {
    configureSpy.mockReturnValue({});
    const validators = getDefaultPasswordValidators();
    expect(validators).toStrictEqual([]);
  });
});

describe('getConfirmPassword', () => {
  it('returns confirm password valiator as expected', () => {
    const validator = getMatchesConfirmPassword('pw');
    expect(validator).toMatchObject({
      validator: expect.any(Function),
      message: 'Your passwords must match',
      validationMode: 'onTouched',
    });
  });

  it('validates to true when passwords match', () => {
    const { validator } = getMatchesConfirmPassword('myPassword');
    const isValid = validator('myPassword');
    expect(isValid).toBe(true);
  });

  it('validates to false when passwords do not match', () => {
    const { validator } = getMatchesConfirmPassword('myPassword');
    const isValid = validator('mismatchingPassword');
    expect(isValid).toBe(false);
  });
});

describe('shouldValidate', () => {
  it('returns true when validationMode is onBlur eventType is blur', () => {
    const result = shouldValidate({
      validationMode: 'onBlur',
      eventType: 'blur',
      hasBlurred: false,
    });

    expect(result).toBe(true);
  });

  it('returns false when validationMode is onBlur and eventType is not blur', () => {
    const result = shouldValidate({
      validationMode: 'onBlur',
      eventType: 'change',
      hasBlurred: false,
    });

    expect(result).toBe(false);
  });

  it('returns true when validationMode is onChange and eventType is change', () => {
    const result = shouldValidate({
      validationMode: 'onChange',
      eventType: 'change',
      hasBlurred: false,
    });

    expect(result).toBe(true);
  });

  it('returns true when validationMode is onTouched and eventType is blur and hasBlurred is false', () => {
    const result = shouldValidate({
      validationMode: 'onTouched',
      eventType: 'blur',
      hasBlurred: false,
    });

    expect(result).toBe(true);
  });

  it('returns false when validationMode is onTouched and eventType is change and hasBlurred is false', () => {
    const result = shouldValidate({
      validationMode: 'onTouched',
      eventType: 'change',
      hasBlurred: false,
    });

    expect(result).toBe(false);
  });
});

describe('runFieldValidators', () => {
  it('returns an empty array when there are no validators', () => {
    const value = 'test';
    const validators = [];
    const eventType = 'blur';
    const hasBlurred = false;
    const result = runFieldValidators({
      value,
      validators,
      eventType,
      hasBlurred,
    });
    expect(result).toEqual([]);
  });

  it('returns an empty array when the value is empty', () => {
    const value = '';
    const validators: ValidatorOptions[] = [
      { validator: () => true, validationMode: 'onBlur', message: 'error' },
    ];
    const eventType = 'blur';
    const hasBlurred = false;
    const result = runFieldValidators({
      value,
      validators,
      eventType,
      hasBlurred,
    });
    expect(result).toEqual([]);
  });

  it('returns an error message when a validator fails', () => {
    const value = 'test';
    const validators: ValidatorOptions[] = [
      {
        validator: (value) => value === 'pass',
        validationMode: 'onBlur',
        message: 'error',
      },
    ];
    const eventType = 'blur';
    const hasBlurred = false;
    const result = runFieldValidators({
      value,
      validators,
      eventType,
      hasBlurred,
    });
    expect(result).toEqual(['error']);
  });

  it('returns an empty array when all validators pass', () => {
    const value = 'pass';
    const validators: ValidatorOptions[] = [
      {
        validator: (value) => value === 'pass',
        validationMode: 'onBlur',
        message: 'error',
      },
    ];
    const eventType = 'blur';
    const hasBlurred = false;
    const result = runFieldValidators({
      value,
      validators,
      eventType,
      hasBlurred,
    });
    expect(result).toEqual([]);
  });

  it('returns previous errors when a validator returns an error message', () => {
    const value = 'test';
    const validators: ValidatorOptions[] = [
      {
        validator: (val: string) => val === 'other',
        validationMode: 'onChange',
        message: 'Value is not correct',
      },
      {
        validator: (val: string) => val.length > 5,
        validationMode: 'onBlur',
        message: 'Value must be longer than 5 characters',
      },
    ];
    const eventType = 'change';
    const hasBlurred = true;

    const result = runFieldValidators({
      value,
      validators,
      eventType,
      hasBlurred,
    });
    expect(result).toEqual(['Value is not correct']);
  });
});

describe('getDefaultConfirmPasswordValidators', () => {
  it('returns an array with one validator', () => {
    const password = 'password';
    const validators = getDefaultConfirmPasswordValidators(password);
    expect(validators).toHaveLength(1);
  });
});
