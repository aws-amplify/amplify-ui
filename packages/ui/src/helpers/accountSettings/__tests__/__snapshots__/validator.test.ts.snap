// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDefaultPasswordValidators returns validators as expected for full Amplify config  1`] = `
Array [
  Object {
    "message": "Password must have at least 8 characters",
    "validationMode": "onTouched",
    "validator": [Function],
  },
  Object {
    "message": "Password must have lower case letters",
    "validationMode": "onTouched",
    "validator": [Function],
  },
  Object {
    "message": "Password must have upper case letters",
    "validationMode": "onTouched",
    "validator": [Function],
  },
  Object {
    "message": "Password must have numbers",
    "validationMode": "onTouched",
    "validator": [Function],
  },
  Object {
    "message": "Password must have special characters",
    "validationMode": "onTouched",
    "validator": [Function],
  },
]
`;

exports[`getDefaultPasswordValidators returns validators as expected for partial Amplify config  1`] = `
Array [
  Object {
    "message": "Password must have lower case letters",
    "validationMode": "onTouched",
    "validator": [Function],
  },
  Object {
    "message": "Password must have upper case letters",
    "validationMode": "onTouched",
    "validator": [Function],
  },
]
`;
